// moduloatv4/index.js
class Cliente {
    #cpf
    constructor(nome, cpf, endereco) {
        this.nome = nome
        this.#cpf = cpf
        this.endereco = endereco
        this.telefones = new Set() 
    }

    get pegarCpf() {
        return this.#cpf
    }

    get upperNome() {
        return this.nome.toUpperCase()
    }

    get lowerNome() {
        return this.nome.toLowerCase()
    }

    get upperCpf() {
        return this.#cpf.toUpperCase()
    }

    get lowerCpf() {
        return this.#cpf.toLowerCase()
    }

    get upperEndereco() {
        return this.endereco?.rua?.toUpperCase()
    }

    get lowerEndereco() {
        return this.endereco?.rua?.toLowerCase()
    }

    addTelefone(telefone) {
        this.telefones.add(telefone) 
    }
}

class Telefone {
    constructor(ddd, numero) {
        this.ddd = ddd
        this.numero = numero
    }
}

class Endereco {
    constructor(estado, cidade, rua, numero) {
        this.estado = estado
        this.cidade = cidade
        this.rua = rua
        this.numero = numero
    }

    get upperEstado() {
        return this.estado.toUpperCase()
    }

    get lowerEstado() {
        return this.estado.toLowerCase()
    }

    get upperCidade() {
        return this.cidade.toUpperCase()
    }

    get lowerCidade() {
        return this.cidade.toLowerCase()
    }

    get upperRua() {
        return this.rua.toUpperCase()
    }

    get lowerRua() {
        return this.rua.toLowerCase()
    }
}

class Empresa {
    #cnpj
    constructor(razaoSocial, nomeFantasia, cnpj, endereco) {
        this.razaoSocial = razaoSocial
        this.nomeFantasia = nomeFantasia
        this.#cnpj = cnpj
        this.endereco = endereco
        this.clientes = new Set()
        this.telefones = new Set()
        this.empresa = new Set()
    }

    get pegarCnpj() {
        return this.#cnpj
    }

    get upperRazaoSocial() {
        return this.razaoSocial.toUpperCase()
    }

    get lowerRazaoSocial() {
        return this.razaoSocial.toLowerCase()
    }

    get upperNomeFantasia() {
        return this.nomeFantasia.toUpperCase()
    }

    get lowerNomeFantasia() {
        return this.nomeFantasia.toLowerCase()
    }

    get upperEndereco() {
        return this.endereco?.rua?.toUpperCase()
    }

    get lowerEndereco() {
        return this.endereco?.rua?.toLowerCase()
    }

    addCliente(cliente) {
        this.clientes.add(cliente)
    }

    addTelefone(telefone) {
        this.telefones.add(telefone)
    }

    addEndereco(endereco) {
        this.empresa.add(endereco)
    }

    detalhe() {
        let descricao = `Razão Social: ${this.razaoSocial}\n`
        descricao += `Nome Fantasia: ${this.nomeFantasia}\n`
        descricao += `----------------------------------------------------------\n`

        this.clientes.forEach(cliente => {
            descricao += `Nome: ${cliente.nome}\n`
            descricao += `Estado: ${cliente.endereco?.estado} Cidade: ${cliente.endereco?.cidade} Rua: ${cliente.endereco?.rua} Número: ${cliente.endereco?.numero}\n`

            cliente.telefones.forEach(telefone => {
                descricao += `DDD: ${telefone.ddd}  Número: ${telefone.numero}\n`
            })

            descricao += `----------------------------------------------------------\n`
        })

        return descricao;
    }
}

export {Cliente, Telefone, Endereco, Empresa}
